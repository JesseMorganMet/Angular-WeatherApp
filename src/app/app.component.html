<h1>{{title}}</h1>

<form (ngSubmit)="submitForm()">
  <mat-form-field>
    <input type="text" matInput [formControl]="myControl" [matAutocomplete]="auto">
    <mat-autocomplete #auto="matAutocomplete">
      <mat-option *ngFor="let location of filteredLocationNames | async" [value]="location.name" >
        {{location.name}}
      </mat-option>
    </mat-autocomplete>
  </mat-form-field>
  <button type="submit">Submit</button>
</form>

<div *ngIf="weatherData">
  <app-big-weather [somethingElse]="myControl.value" [data]="weatherData.SiteRep.DV.Location.Period[0]"></app-big-weather>
</div>

<div id="weatherComp" *ngIf="weatherData">
  <app-weather  *ngFor="let day of weatherData.SiteRep.DV.Location.Period" [data]="weatherData"> </app-weather>
</div>
<!--<app-test *ngFor="let day of days></app-test>-->
<!--loop through all days -->
