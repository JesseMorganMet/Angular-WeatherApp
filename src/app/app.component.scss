#weatherComp{
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  margin:0;
}
#searchBoy{
  display: flex;
  justify-content: center;
  align-items: center;
  height: 60px;
  background-color: #4c617b;
}
mat-form-field{
  width:50vw;
}
\mo-header {
  p {
    color: white;
    &:hover {
      background-color: lighten(#2a2a2a, 10%);
    }
  }
}
\mo-footer{
  position: fixed;
  left: 0;
  bottom: 0;
  width: 100%;
}

.weatherContent{
  padding:20px;
  display: flex;
  flex-direction: column;
  align-items: center;
}
