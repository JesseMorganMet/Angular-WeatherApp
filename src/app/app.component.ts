import {Component} from '@angular/core';
import {WeatherService} from './weather.service';
import {FormControl} from '@angular/forms';
import {map, Observable, startWith} from 'rxjs';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})

export class AppComponent {

  title = 'Weather App';
  locationNames: any = [];
  index: number = 0;
  days
  myControl: FormControl = new FormControl();
  filteredLocationNames: Observable<any[]>;
  weatherData: any;

  constructor(private service: WeatherService) {
  }
//just testing
  ngOnInit() {
    this.locationFunctions()
  }

  locationFunctions() {
    this.service.getLocation().subscribe((data: any) => {
      this.locationsSearch(data);
      this.locationSort(this.locationNames);
      this.locationFilter(this.locationNames);
    })
  }

  locationsSearch(data) {
    for (let i = 0; i < data.Locations.Location.length; i++) {
      const location = data.Locations.Location[i];
      this.locationNames.push({name: location.name, id: location.id});
    }
    return this.locationNames;
  }

  locationSort(data) {
    data.sort(function (a, b) {
      var nameA = a.name.toUpperCase();
      var nameB = b.name.toUpperCase();
      if (nameA < nameB) {
        return -1;
      }
      if (nameA > nameB) {
        return 1;
      }
      return 0;
    });
    return data;
  }

  locationFilter(data) {
    this.filteredLocationNames = this.myControl.valueChanges
      .pipe(
        startWith(''),
        map(val => this.filter(val))
      );
    return this.filteredLocationNames;
  }

  filter(value: string): string[] {
      const filterValue = value.toLowerCase();
      return this.locationNames.filter(location => {
        return location.name.toLowerCase().includes(filterValue)
      });
  }

  submitForm() {
    let location: any = this.filter(this.myControl.value);
    // Fix needed
    // location[0] always selects the top of the search
    // need to make the search go off with the selected option
    // i.e. e1, e2, e3, e4, e5. these are location options with e1 at the top
    // you select e5, e5 appears in the search bar however location[0] === e1
    // so e1 location data is retrieved
    this.service.getData(location[0].id).subscribe((data: any) => {
      this.weatherData = data;
      return this.weatherData;
    });
  }
}
