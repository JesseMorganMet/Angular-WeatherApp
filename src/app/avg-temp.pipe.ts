import { Pipe, PipeTransform } from '@angular/core';

@Pipe({name: 'avgTemp'})
export class AvgTempPipe implements PipeTransform {
  transform(data: any): string {
    let tempAvg = 0;
    for(let i =0; i < data.length; i++){
      let tempIndv = 0;
      tempIndv = parseInt(data[i].T);
      tempAvg = tempAvg + tempIndv;
    }
    return (tempAvg/data.length).toFixed(0) + "C";
  }
}

@Pipe({name: 'avgPrecip'})
export class AvgPrecipPipe implements PipeTransform {
  transform(data: any): string {
    let ppAvg = 0;
    for(let i =0; i < data.length; i++){
      let ppIndv = 0;
      ppIndv = parseInt(data[i].Pp);
      ppAvg = ppAvg + ppIndv;
    }
    let ppRound = Math.round((ppAvg/data.length)/5)*5;
    if(ppRound >= 10){
      ppRound = Math.round(ppRound/10)*10;
    }
    return (ppRound + "%");
  }
}

@Pipe({name: 'weatherType'})
export class WeatherTypePipe implements PipeTransform {
  transform(data: any): any {
    let weatherDaily = parseInt(data[0].W);
    if (weatherDaily===1|| weatherDaily===0) {
      return "../assets/icons/Sun.png"
    } else if(weatherDaily===2||weatherDaily===3||weatherDaily===7||weatherDaily===8){
      return "../assets/icons/Cloud.png"
    } else if(weatherDaily>=4&&weatherDaily<=6){
      return "../assets/icons/Fog.png"
    } else if(weatherDaily>=9&&weatherDaily<=15){
      return "../assets/icons/Rain.png"
    } else if(weatherDaily>=16&&weatherDaily<=27){
      return "../assets/icons/Snow.png"
    } else if(weatherDaily>=28&&weatherDaily<=30){
      return "../assets/icons/Thunder.png"
    } else{
      console.log("Something went wrong");
      console.log(weatherDaily);
    }
  }
}

@Pipe({name: 'weatherTypeAlt'})
export class WeatherTypeAltPipe implements PipeTransform {
  transform(data: any): any {
    let weatherDaily = parseInt(data[0].W);
    if (weatherDaily===1|| weatherDaily===0) {
      return "Sunny icon"
    } else if(weatherDaily===2||weatherDaily===3||weatherDaily===7||weatherDaily===8){
      return "Cloudy icon"
    } else if(weatherDaily>=4&&weatherDaily<=6){
      return "Foggy icon"
    } else if(weatherDaily>=9&&weatherDaily<=15){
      return "Raining icon"
    } else if(weatherDaily>=16&&weatherDaily<=27){
      return "Snowing icon"
    } else if(weatherDaily>=28&&weatherDaily<=30){
      return "Thunder icon"
    } else{
      console.log("Something went wrong");
      console.log(weatherDaily);
    }
  }
}

@Pipe({name: 'wType'})
export class WTypePipe implements PipeTransform {
  transform(data: any): any {
    let weather = data[0].W
    const weatherType = ["Clear night", "Sunny day", "Partly cloudy (night)", "Partly cloudy (day)", "Not used", "Mist", "Fog", "Cloudy", "Overcast",
      "Light rain shower (night)", "Light rain shower (day)", "Drizzle", "Light rain", "Heavy rain shower (night)", "Heavy rain shower (day)",
      "Heavy rain", "Sleet shower (night)", "Sleet shower (day)", "Sleet", "Hail shower (night)", "Hail shower (day)", "Hail", "Light snow shower (night)",
      "Light snow shower (day)", "Light snow", "Heavy snow shower (night)", "Heavy snow shower (day)", "Heavy snow", "Thunder shower (night)", "Thunder shower (day)", "Thunder"
    ];
    return weatherType[weather];
  }
}

@Pipe({name: 'uv'})
export class uvPipe implements PipeTransform {
  transform(data: any): any {
    const uv = data[0].U;
    if (uv <= 2) {
      return "L";
    } else if (uv <= 5) {
      return "M";
    } else if (uv <= 7) {
      return "H";
    } else if (uv <= 10) {
      return "VH";
    } else if (uv >= 11) {
      return "E";
    } else {
      console.log("Something went wrong");
    }
  }
}


@Pipe({name: 'minTemp'})
export class minTempPipe implements PipeTransform {
  transform(data: any): any {
    const tempArr = [];
    for (let j = 0; j < data.length; j++) {
      tempArr.push(data[j].T);
    }
    return Math.min(...tempArr);
  }
}

  @Pipe({name: 'maxTemp'})
  export class maxTempPipe implements PipeTransform {
    transform(data: any): any {
      const tempArr = [];
      for (let j = 0; j < data.length; j++) {
        tempArr.push(data[j].T);
      }
      return Math.max(...tempArr);
    }
  }
