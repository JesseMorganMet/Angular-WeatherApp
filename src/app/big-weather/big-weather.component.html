<div id="bigBit">
  <h2>{{placeName}}</h2>
  <img src= "{{data.Rep | weatherType}}" alt= "{{data.Rep | weatherTypeAlt}}">
  <h2>{{data.Rep | wType}}</h2>
</div>

<!--collums-->
<div id="dataText">
<!--  rows-->
  <div>
    <h2>High's Of</h2>
    <h2>{{data.Rep | maxTemp}}&deg;C</h2>
  </div>
  <div>
    <h2>Lows Of</h2>
    <h2>{{data.Rep | minTemp}}&deg;C</h2>
  </div>
  <div>
    <h2>UV</h2>
    <h2>{{data.Rep | uv}}</h2>
  </div>
</div>
