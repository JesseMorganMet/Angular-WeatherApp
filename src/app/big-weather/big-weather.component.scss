#dataText{
  width:100%;
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;
}
#dataTextIn{
  margin:0.5%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
#bigBit{
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

}
#bigBit > img{
  max-width: 400px;
  margin: -5% 0 -5% 0;
}
