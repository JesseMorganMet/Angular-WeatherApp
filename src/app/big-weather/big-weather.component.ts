import {Component, Input, OnInit} from '@angular/core';

@Component({
  selector: 'app-big-weather',
  templateUrl: './big-weather.component.html',
  styleUrls: ['./big-weather.component.scss']
})
export class BigWeatherComponent implements OnInit {

  @Input() placeName;
  @Input() data;

  constructor() { }

  ngOnInit(): void {
  }
}
