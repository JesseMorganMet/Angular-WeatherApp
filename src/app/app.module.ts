import {CUSTOM_ELEMENTS_SCHEMA, NgModule} from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';
import {WeatherService} from './weather.service';
import {HttpClientModule} from '@angular/common/http';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import {MatAutocompleteModule} from '@angular/material/autocomplete';
import {MatFormFieldModule} from '@angular/material/form-field';
import {FormsModule, ReactiveFormsModule} from '@angular/forms';
import {MatInputModule} from '@angular/material/input';
import { TestComponent } from './test/test.component';
import { WeatherComponent } from './weather/weather.component';
import { BigWeatherComponent } from './big-weather/big-weather.component';
import {
  AvgTempPipe,
  WeatherTypePipe,
  AvgPrecipPipe,
  WeatherTypeAltPipe,
  WTypePipe,
  uvPipe,
  minTempPipe,
  maxTempPipe
} from './avg-temp.pipe';
import { defineCustomElements as ui } from '@mo/ui/loader';
ui();

@NgModule({
  declarations: [
    AppComponent,
    TestComponent,
    WeatherComponent,
    BigWeatherComponent
  ],
  imports: [
    BrowserModule,
    HttpClientModule,
    BrowserAnimationsModule,
    MatAutocompleteModule,
    MatFormFieldModule,
    ReactiveFormsModule,
    MatInputModule,
    FormsModule,
    WeatherTypePipe,
    WeatherTypeAltPipe,
    AvgPrecipPipe,
    AvgTempPipe,
    WTypePipe,
    maxTempPipe,
    minTempPipe,
    uvPipe
  ],
  providers: [
    WeatherService
  ],
  bootstrap: [AppComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class AppModule { }
