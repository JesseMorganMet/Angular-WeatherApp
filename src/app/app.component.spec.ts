import {ComponentFixture, TestBed} from '@angular/core/testing';
import { AppComponent } from './app.component';
import {BigWeatherComponent} from './big-weather/big-weather.component';
import {WeatherComponent} from './weather/weather.component';
import {FormControl} from '@angular/forms';
import { MatFormFieldModule} from '@angular/material/form-field';
import {WeatherService} from './weather.service';
import {HttpClientTestingModule, HttpTestingController} from '@angular/common/http/testing';
import {MatAutocompleteModule} from '@angular/material/autocomplete';

describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;
  let service: WeatherService;
  let httpTestingController: HttpTestingController;

  let fakeData: any = "Taunton";

  let fakeLocal: any = {"Locations":{"Location":[{"name": "Taunton", "id": "324072"}]}};

  let fakeLocalData: any =  {
    "SiteRep": {
      "DV": {
        "dataDate": "2022-08-15T09:00:00Z",
        "Location": {
          "i": "324072",
          "Period": [
            {
              "value": "2022-08-15Z",
              "Rep": [
                {
                  "Pp": "29",
                  "T": "17",
                  "W": "7",
                  "U": "1",
                },
                {
                  "Pp": "82",
                  "T": "18",
                  "W": "15",
                  "U": "2",
                },
                {
                  "Pp": "82",
                  "T": "20",
                  "W": "12",
                  "U": "4",
                },
                {
                  "Pp": "54",
                  "T": "23",
                  "W": "30",
                  "U": "3",
                },
                {
                  "Pp": "8",
                  "T": "21",
                  "W": "7",
                  "U": "1",
                },
                {
                  "Pp": "8",
                  "T": "20",
                  "W": "7",
                  "U": "0",
                }
              ]
            },
            {
              "value": "2022-08-16Z",
              "Rep": [
                {
                  "Pp": "57",
                  "T": "19",
                  "W": "30",
                  "U": "0",
                },
                {
                  "Pp": "8",
                  "T": "18",
                  "W": "7",
                  "U": "0",
                },
                {
                  "Pp": "57",
                  "T": "18",
                  "W": "30",
                  "U": "1",
                },
                {
                  "Pp": "58",
                  "T": "19",
                  "W": "30",
                  "U": "3",
                },
                {
                  "Pp": "14",
                  "T": "21",
                  "W": "7",
                  "U": "5",
                },
                {
                  "Pp": "77",
                  "T": "20",
                  "W": "30",
                  "U": "3",
                },
                {
                  "Pp": "61",
                  "T": "19",
                  "W": "29",
                  "U": "1",
                },
                {
                  "Pp": "58",
                  "T": "18",
                  "W": "30",
                  "U": "0",
                }
              ]
            },
            {
              "value": "2022-08-17Z",
              "Rep": [
                {
                  "Pp": "11",
                  "T": "18",
                  "W": "7",
                  "U": "0",
                },
                {
                  "Pp": "14",
                  "T": "17",
                  "W": "7",
                  "U": "0",
                },
                {
                  "Pp": "13",
                  "T": "17",
                  "W": "7",
                  "U": "1",
                },
                {
                  "Pp": "16",
                  "T": "18",
                  "W": "8",
                  "U": "3",
                },
                {
                  "Pp": "16",
                  "T": "20",
                  "W": "7",
                  "U": "5",
                },
                {
                  "Pp": "33",
                  "T": "21",
                  "W": "10",
                  "U": "3",
                },
                {
                  "Pp": "13",
                  "T": "21",
                  "W": "7",
                  "U": "1",
                },
                {
                  "Pp": "6",
                  "T": "20",
                  "W": "2",
                  "U": "0",
                }
              ]
            },
            {
              "value": "2022-08-18Z",
              "Rep": [
                {
                  "Pp": "6",
                  "T": "18",
                  "W": "7",
                  "U": "0",
                },
                {
                  "Pp": "2",
                  "T": "17",
                  "W": "2",
                  "U": "0",
                },
                {
                  "Pp": "3",
                  "T": "16",
                  "W": "3",
                  "U": "1",
                },
                {
                  "Pp": "3",
                  "T": "18",
                  "W": "3",
                  "U": "3",
                },
                {
                  "Pp": "3",
                  "T": "20",
                  "W": "3",
                  "U": "6",
                },
                {
                  "Pp": "3",
                  "T": "20",
                  "W": "3",
                  "U": "4",
                },
                {
                  "Pp": "5",
                  "T": "20",
                  "W": "3",
                  "U": "1",
                },
                {
                  "Pp": "11",
                  "T": "19",
                  "W": "8",
                  "U": "0",
                }
              ]
            },
            {
              "value": "2022-08-19Z",
              "Rep": [
                {
                  "Pp": "12",
                  "T": "19",
                  "W": "7",
                  "U": "0",
                },
                {
                  "Pp": "47",
                  "T": "19",
                  "W": "12",
                  "U": "0",
                },
                {
                  "Pp": "37",
                  "T": "19",
                  "W": "10",
                  "U": "1",
                },
                {
                  "Pp": "13",
                  "T": "18",
                  "W": "3",
                  "U": "3",
                },
                {
                  "Pp": "13",
                  "T": "19",
                  "W": "1",
                  "U": "6",
                },
                {
                  "Pp": "9",
                  "T": "19",
                  "W": "1",
                  "U": "3",
                },
                {
                  "Pp": "4",
                  "T": "18",
                  "W": "1",
                  "U": "1"
                },
                {
                  "Pp": "5",
                  "T": "18",
                  "W": "2",
                  "U": "0"
                }
              ]
            }
          ]
        }
      }
    }
  }

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [
        AppComponent,
        BigWeatherComponent,
        WeatherComponent
      ],
      imports: [
        MatFormFieldModule,
        HttpClientTestingModule,
        MatAutocompleteModule
      ],
      providers:[
        WeatherService
      ]
    }).compileComponents();
  });

  beforeEach(() => {
    service = TestBed.inject(WeatherService);
    httpTestingController = TestBed.inject(HttpTestingController);
    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
  })

  it('should create the app', () => {
    component.myControl.patchValue(fakeData)
    component.locationNames = fakeLocal
    component.submitForm();
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot(); //not yet working

  });
});
