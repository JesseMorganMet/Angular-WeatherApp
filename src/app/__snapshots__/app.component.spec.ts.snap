// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should create the app 1`] = `
<app-root
  __ngContext__={[Function LRootView]}
  index="0"
  locationNames={[Function Array]}
  myControl={[Function FormControl]}
  service={[Function WeatherService]}
  title={[Function String]}
>
  <div
    id="searchBoy"
  >
    <form>
      <mat-form-field
        class="mat-form-field mat-primary"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-infix"
            >
              <input
                class="mat-autocomplete-trigger"
                matinput=""
                type="text"
              />
              <mat-autocomplete
                class="mat-autocomplete"
              >
              </mat-autocomplete>
              <span
                class="mat-form-field-label-wrapper"
              >
              </span>
            </div>
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
          </div>
        </div>
      </mat-form-field>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</app-root>
`;
